pkg/virt-launcher/virtwrap/network/common.go

```diff
@@ -0,0 +1,315 @@
+		if err = DHCPServer(
+			nic.MAC,
+			nic.IP.IP,
+			nic.IP.Mask,
+			api.DefaultBridgeName,
+			serverAddr.IP,
+			nic.Gateway,
+			nameservers,
+			nic.Routes,
+			searchDomains,
+			nic.Mtu,
+		); err != nil {
+			log.Log.Errorf("failed to run DHCP: %v", err)
+			panic(err)
+		}
+	}()
+}
+
+// Allow mocking for tests
+var SetupPodNetwork = SetupNetworkInterfaces
+var DHCPServer = dhcp.SingleClientDHCPServer
+
+func initHandler() {
+	if Handler == nil {
+		Handler = &NetworkUtilsHandler{}
+	}
+}
+
+func setCachedInterface(ifconf *api.Interface) error {
+	buf, err := json.MarshalIndent(&ifconf, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshaling interface cache: %v", err)
+	}
+	err = ioutil.WriteFile(interfaceCacheFile, buf, 0644)
+	if err != nil {
+		return fmt.Errorf("error writing interface cache %v", err)
+	}
+	return nil
+}
+
+func getCachedInterface() (*api.Interface, error) {
+	buf, err := ioutil.ReadFile(interfaceCacheFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	ifconf := api.Interface{}
+	err = json.Unmarshal(buf, &ifconf)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling interface: %v", err)
+	}
+	return &ifconf, nil
+}
+
+// filter out irrelevant routes
+func filterPodNetworkRoutes(routes []netlink.Route, nic *VIF) (filteredRoutes []netlink.Route) {
+	for _, route := range routes {
+		// don't create empty static routes
+		if route.Dst == nil && route.Src.Equal(nil) && route.Gw.Equal(nil) {
+			continue
+		}
+
+		// don't create static route for src == nic
+		if route.Src != nil && route.Src.Equal(nic.IP.IP) {
+			continue
+		}
+
+		filteredRoutes = append(filteredRoutes, route)
+	}
+	return
+}
+
+// only used by unit test suite
+func setInterfaceCacheFile(path string) {
+	interfaceCacheFile = path
+}
+
+// returns nameservers [][]byte, searchdomains []string, error
+func getResolvConfDetailsFromPod() ([][]byte, []string, error) {
+	b, err := ioutil.ReadFile(resolvConf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	nameservers, err := ParseNameservers(string(b))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	searchDomains, err := ParseSearchDomains(string(b))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Log.Reason(err).Infof("Found nameservers in %s: %s", resolvConf, bytes.Join(nameservers, []byte{' '}))
+	log.Log.Reason(err).Infof("Found search domains in %s: %s", resolvConf, strings.Join(searchDomains, " "))
+
+	return nameservers, searchDomains, err
+}
+
+func ParseNameservers(content string) ([][]byte, error) {
+	var nameservers [][]byte
+
+	re, err := regexp.Compile("([0-9]{1,3}.?){4}")
+	if err != nil {
+		return nameservers, err
+	}
+
+	scanner := bufio.NewScanner(strings.NewReader(content))
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, nameserverPrefix) {
+			nameserver := re.FindString(line)
+			if nameserver != "" {
+				nameservers = append(nameservers, net.ParseIP(nameserver).To4())
```

Comment:
In a pod config in 2018, it might be reasonable to assume in general that the addresses listed will be IPv4, but I think there might be a better forward-looking approach. This line in particular is more likely than others to suffer from issues even if we're not truly ready to support IPv6 yet.

Reply:
Upon re-reading my comment, I'm not sure I was completely clear about what my concern is: When I look at the `/etc/resolv.conf` on my system there's three "nameserver" lines. One is IPv4 and two are IPv6. The same thing can and will happen in clusters/pods where KubeVirt is deployed. Since we're consuming the contents of `/etc/resolv.conf` (and the addresses contained there come from sources beyond the scope of KubeVirt) we can't assume IPv4-only addressing schemes. Ideally we'd use IPv6 addresses found here, but at a minimum this block needs to at least check for and discard them.